sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessorClient.cs

```diff
@@ -0,0 +1,982 @@
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.Linq.Expressions;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Messaging.ServiceBus.Core;
+using Azure.Messaging.ServiceBus.Diagnostics;
+using Microsoft.Azure.Amqp;
+
+namespace Azure.Messaging.ServiceBus
+{
+    /// <summary>
+    ///   A client responsible for reading <see cref="ServiceBusMessage" /> from a specific entity
+    ///   as a member of a specific consumer group.
+    ///
+    ///   A consumer may be exclusive, which asserts ownership over associated partitions for the consumer
+    ///   group to ensure that only one consumer from that group is reading the from the partition.
+    ///   These exclusive consumers are sometimes referred to as "Epoch Consumers."
+    ///
+    ///   A consumer may also be non-exclusive, allowing multiple consumers from the same consumer
+    ///   group to be actively reading events from a given partition.  These non-exclusive consumers are
+    ///   sometimes referred to as "Non-Epoch Consumers."
+    /// </summary>
+    ///
+    public class ServiceBusProcessorClient : IAsyncDisposable
+    {
+        private Func<ServiceBusMessage, ServiceBusSession, Task> _processMessage;
+        private Func<ExceptionReceivedEventArgs, Task> _processErrorAsync = default;
+        /// <summary>The primitive for synchronizing access during start and set handler operations.</summary>
+        private readonly object EventHandlerGuard = new object();
+        /// <summary>The primitive for synchronizing access during start and close operations.</summary>
+        private SemaphoreSlim MessageHandlerSemaphore;
+        /// <summary>The primitive for synchronizing access during start and close operations.</summary>
+        private readonly SemaphoreSlim ProcessingStartStopSemaphore = new SemaphoreSlim(1, 1);
+        private CancellationTokenSource RunningTaskTokenSource { get; set; }
+
+        private readonly ServiceBusReceiverClient _receiver;
+
+        /// <summary>
+        ///   The running task responsible for performing partition load balancing between multiple <see cref="ServiceBusProcessorClient" />
+        ///   instances, as well as managing partition processing tasks and ownership.
+        /// </summary>
+        ///
+        private Task ActiveReceiveTask { get; set; }
+
+        /// <summary>
+        ///   Called when a 'process message' event is triggered.
+        /// </summary>
+        ///
+        /// <param name="message">The set of arguments to identify the context of the event to be processed.</param>
+        /// <param name="session"></param>
+        ///
+        internal Task OnProcessMessageAsync(ServiceBusMessage message, ServiceBusSession session) => _processMessage(message, session);
```

Comment:
This is `internal` but the equivalent error-related method is `private`.   Should these keep the same visibility?   Is there a scenario where it is desirable for internal types to call this directly?

Reply:
both private